flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,614 @@
+        closestCentroidLabel = centroid.label
+      }
+    })
+    (closestCentroidLabel, minDistance)
+  }
+
+  /**
+   * Returns the initial centroids for the KMeans algorithm based upon the information in
+   * parameter
+   *
+   * @param data The training data set
+   * @param parameter Parameter Map containing user parameters
+   * @return Initial centroids for KMeans clustering
+   */
+  private def init(data: DataSet[Vector], parameter: ParameterMap): DataSet[Seq[LabeledVector]] = {
+    parameter.get(InitialCentroids) match {
+      case Some(value) => data.getExecutionEnvironment.fromElements(value)
+      case None => {
+
+        val k = parameter.get(NumClusters) match{
+          case Some(value) => value
+          case None => throw new RuntimeException("Specify the number of clusters.")
+        }
+        val l = parameter.get(OversamplingFactor) match{
+          case Some(value) => value
+          case None => 2 * k  // default value
+        }
+        val r = parameter.get(KMeansParRounds).get
+
+        val blocks = data.getParallelism
+
+        parameter.get(InitialStrategy) match {
+          case Some("random") => {
+            random(data.map(x => (x,1)), k)
+          }
+          case Some("kmeans++") => {
+            kmeans(data.map(x => (x,1)), k, blocks)
+          }
+          case Some("kmeans||") => {
+            parInit(data, k, blocks, l ,r)
+          }
+          case default => {
+            throw new RuntimeException("Specify a valid initialization strategy.")
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Pick k centers from data one by one using kmeans|| initialization scheme
+   *
+   * The k-means|| algorithm works as described by the original authors
+   * (http://theory.stanford.edu/~sergei/papers/vldb12-kmpar.pdf):
+   *
+   * Given a data set X with |X| points, the k-means|| algorithm proceeds as follows:
+   *
+   * 1. Initialize C \leftarrow \{\}
+   * 2. Let p be a point sampled uniformly at random from X. C \leftarrow C \cup \{p\}
+   * 3. for i \leftarrow 1 to r
+   * Let C' be the set of formed by independently sampling every point x in X with probability
+   * \ell\cdot\frac{d(x,C)}{sigma_nolimits{p \in X }d(p,C)}
+   * C \leftarrow C \cup C'
+   * 4. Assign weights to all point c in C as the number of points from X which are closest to c
+   * 5. Run kmeans++ initialization on the weighted set C and return k centers
+   *
+   * @param data Training data set
+   * @param k Number of clusters
+   * @param blocks Blocks in the data
+   * @param oversampling Oversampling rate (\ell)
+   * @param rounds Number of rounds (r)
+   * @return Initial centroids
+   */
+  private def parInit(
+      data: DataSet[Vector],
+      k: Int,
+      blocks: Int,
+      oversampling: Double,
+      rounds: Int)
+    : DataSet[Seq[LabeledVector]] = {
+    // first pick one center randomly
+    val oversamplingFactor = data.getExecutionEnvironment.fromElements(oversampling)
```

Comment:
We can provide the oversamplingFactor as part of the closure. There is no need to pass it as a broadcast variable.